[CPS] Move project picker to navigation and use real data for getting linked projects

mbondyra opened the pull request:
## Summary

Closes [#233789](https://github.com/elastic/kibana/issues/233789).
Closes #238511.

Refactors the project picker into a new shared package (`@kbn/cps-utils`) to enable reusability across multiple consumers, and uses the real data to get Linked projects to the picker.
Moves project picker to navigation - currently it is visible for all the Kibana pages when CPS is on. Since this feature is under the .yml setting, it's not visible to the users so we can calmly wait for the designs to implement the disabled state.
<img width="1164" height="453" alt="Screenshot 2025-11-11 at 13 42 17" src="https://github.com/user-attachments/assets/a86ea5fa-a958-4582-aea4-9d929c6db77b" />

## Changes

### Architecture
- **New Package**: Created `@kbn/cps-utils` package containing the project picker component and related utilities
- **CPS Manager**: Added `cpsManager` service in the CPS plugin that fetches linked projects when running in serverless + CPS mode AND manages projectRouting state.
- **Project Routing**: Extended `@kbn/es-query` with project routing types (`ProjectRouting`)

### Project Picker Enhancements
- **Consumer-Driven Routing**: Project picker now accepts `projectRouting` and `onProjectRoutingChange` props, allowing consumers (e.g., Dashboard, Lens layers, global) to manage routing state
- **Conditional Rendering**: Picker only displays when CPSManager returns valid data with at least one linked project
- **Testing**: Added unit tests and integration tests for the project picker component

### Configuration Changes
- **Feature Flag Migration**: Removed localStorage-based feature flag in favor of YAML configuration
- **New Config Requirements**: To enable CPS mode, add the following to your `serverless.yml`:
  ```yaml
  cps.enabled: true
  cps.cpsEnabled: true
  ```

### How to test it
- To set up CPS, make sure to use the internal doc we use for this purpose.

For now, the data in the picker should reflect what has been set up, but picker is not reactive.

mbondyra committed:
move to package

kibanamachine committed:
Changes from node scripts/eslint_all_files --no-cache --fix

mbondyra committed:
do not import schema to client side code

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

mbondyra committed:
Merge branch 'main' into data-plugin-cps-3

mbondyra committed:
Merge branch 'main' into data-plugin-cps-3

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/357432)
* Commit: 970265a1a312dfecde8b403db4d8a593913bd388

### Failed CI Steps
* [FTR Configs #26](https://buildkite.com/elastic/kibana-pull-request/builds/357432#019a45e7-34e6-49a8-a9c8-29a495fb02eb)
* [FTR Configs #26](https://buildkite.com/elastic/kibana-pull-request/builds/357432#019a4657-a176-4404-8dee-8029cc754584)

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/de217ef5b84c812b4c7c7848405fdf418d50f988) | [after](https://github.com/elastic/kibana/commit/970265a1a312dfecde8b403db4d8a593913bd388) | diff |
| --- | --- | --- | --- |
| `cps` | 4 | 5 | +1 |
| `unifiedSearch` | 508 | 510 | +2 |
| total |  |  | +3 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/de217ef5b84c812b4c7c7848405fdf418d50f988) | [after](https://github.com/elastic/kibana/commit/970265a1a312dfecde8b403db4d8a593913bd388) | diff |
| --- | --- | --- | --- |
| `@kbn/cps-utils` | - | 7 | +7 |
| `@kbn/es-query` | 240 | 241 | +1 |
| `cps` | 6 | 27 | +21 |
| `unifiedSearch` | 119 | 120 | +1 |
| total |  |  | +30 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/de217ef5b84c812b4c7c7848405fdf418d50f988) | [after](https://github.com/elastic/kibana/commit/970265a1a312dfecde8b403db4d8a593913bd388) | diff |
| --- | --- | --- | --- |
| `dashboard` | 658.4KB | 658.5KB | +75.0B |
| `unifiedSearch` | 403.3KB | 401.6KB | -1.7KB |
| total |  |  | -1.6KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/de217ef5b84c812b4c7c7848405fdf418d50f988) | [after](https://github.com/elastic/kibana/commit/970265a1a312dfecde8b403db4d8a593913bd388) | diff |
| --- | --- | --- | --- |
| `cps` | 1.4KB | 2.5KB | +1.1KB |
| `unifiedSearch` | 22.6KB | 22.8KB | +155.0B |
| total |  |  | +1.3KB |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/de217ef5b84c812b4c7c7848405fdf418d50f988) | [after](https://github.com/elastic/kibana/commit/970265a1a312dfecde8b403db4d8a593913bd388) | diff |
| --- | --- | --- | --- |
| `@kbn/cps-utils` | - | 8 | +8 |
| `@kbn/es-query` | 303 | 305 | +2 |
| `cps` | 6 | 27 | +21 |
| `unifiedSearch` | 157 | 158 | +1 |
| total |  |  | +32 |

#### async chunk count
| id | [before](https://github.com/elastic/kibana/commit/de217ef5b84c812b4c7c7848405fdf418d50f988) | [after](https://github.com/elastic/kibana/commit/970265a1a312dfecde8b403db4d8a593913bd388) | diff |
| --- | --- | --- | --- |
| `unifiedSearch` | 4 | 5 | +1 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/de217ef5b84c812b4c7c7848405fdf418d50f988) | [after](https://github.com/elastic/kibana/commit/970265a1a312dfecde8b403db4d8a593913bd388) | diff |
| --- | --- | --- | --- |
| `@kbn/cps-utils` | - | 1 | +1 |
| `cps` | 4 | 5 | +1 |
| total |  |  | +2 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/de217ef5b84c812b4c7c7848405fdf418d50f988) | [after](https://github.com/elastic/kibana/commit/970265a1a312dfecde8b403db4d8a593913bd388) | diff |
| --- | --- | --- | --- |
| `@kbn/cps-utils` | - | 1 | +1 |
| `cps` | 4 | 5 | +1 |
| total |  |  | +2 |


</details>


### History
* :yellow_heart: [Build #357188](https://buildkite.com/elastic/kibana-pull-request/builds/357188) was flaky 4bb6b4777358dc4487cebf847b030cc4889b72f5

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357432","number":357432,"commit":"970265a1a312dfecde8b403db4d8a593913bd388"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357188","number":357188,"commit":"4bb6b4777358dc4487cebf847b030cc4889b72f5"}],"number":357432}
buildkite-pr-comment-->

jbudz reviewed (approved):
`packages/kbn-optimizer/limits.yml` LGTM

nickpeihl commented on src/platform/plugins/shared/unified_search/public/plugin.ts:
```diff
@@ -62,7 +62,7 @@ export class UnifiedSearchPublicPlugin
   public start(
     core: CoreStart,
-    { data, dataViews, uiActions, screenshotMode }: UnifiedSearchStartDependencies
+    { data, dataViews, uiActions, screenshotMode, cps }: UnifiedSearchStartDependencies
```
Nit. Can we assign a more meaningful variable name in this destructuring?

```suggestion
    { data, dataViews, uiActions, screenshotMode, cps: crossProjectSearch }: UnifiedSearchStartDependencies
```

cqliu1 commented on src/platform/plugins/shared/unified_search/public/query_string_input/query_bar_top_row.tsx:
```diff
@@ -949,7 +937,15 @@ export const QueryBarTopRow = React.memo(
+                  wrappingContainer={(children: React.ReactNode) => (
+                    <EuiFlexItem grow={false}>{children}</EuiFlexItem>
+                  )}
+                />
```
Is it necessary to pass this `wrappingContainer` prop to the `ProjectPicker`? Can we wrap the component here with the `EuiFlexItem` and remove this `wrappingContainer` prop from the Project Picker component?

```suggestion
                <EuiFlexItem grow={false}>
                  <ProjectPicker
                    projectRouting={props.projectRouting}
                    onProjectRoutingChange={props.onProjectRoutingChange}
                  />
                </EuiFlexItem>
```

cqliu1 reviewed (commented):
I tested it locally with linked projects data, and it fetches the project data correctly. Code overall looks good. I just have a few comments.

lukasolson reviewed (commented):
LGTM (left a question below)... Do we have any documentation yet from ES about how this `ProjectRouting` parameter will be used in queries, or is this still being fleshed out?

mbondyra committed:
Merge commit '8d77e6c618c8d985282dc349a0304ddf631f053d' into data-plugin-cps-3

mbondyra committed:
CR

mbondyra commented on src/platform/packages/shared/kbn-cps-utils/index.ts:
```diff
@@ -8,21 +8,12 @@
- * be used directly by consumers and will load the `ProjectPickerLazy` component lazily with
- * a predefined fallback and error boundary.
- */
 export const ProjectPicker = withSuspense(ProjectPickerLazy);
```
https://github.com/elastic/kibana/pull/241264/commits/8f16f442eadaba6b6cc1530208e007c412ddd062

mbondyra commented on src/platform/packages/shared/kbn-es-query/src/project_routing/index.ts:
```diff
@@ -0,0 +1,21 @@
+ *
+ * @public
+ */
+export type ProjectRouting = '_alias:_origin' | undefined;
```
For now it's just these two and yes, undefined is `_alias:*` - later it will be more, so we'll have to extend this type. Yes, we have a doc - please ping me on Slack and I'll share it :)

elasticmachine commented:
## :yellow_heart: Build succeeded, but was flaky
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/358603)
* Commit: 8f16f442eadaba6b6cc1530208e007c412ddd062

### Failed CI Steps
* [FTR Configs #20](https://buildkite.com/elastic/kibana-pull-request/builds/358603#019a5484-a352-4985-9f18-d8dfd799dc5a)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/358603#019a5484-a352-4985-9f18-d8dfd799dc5a) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/358603/jobs/019a5484-a352-4985-9f18-d8dfd799dc5a/artifacts/019a54ab-ae9d-47ed-a5c3-226b5c8d295b) FTR Configs #<span></span>20 / Core Analysis - Entity Store @<span></span>ess @<span></span>skipInServerlessMKI Entity Store APIs get and list "before all" hook in "get and list"

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/8d77e6c618c8d985282dc349a0304ddf631f053d) | [after](https://github.com/elastic/kibana/commit/8f16f442eadaba6b6cc1530208e007c412ddd062) | diff |
| --- | --- | --- | --- |
| `cps` | 4 | 5 | +1 |
| `unifiedSearch` | 532 | 534 | +2 |
| total |  |  | +3 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/8d77e6c618c8d985282dc349a0304ddf631f053d) | [after](https://github.com/elastic/kibana/commit/8f16f442eadaba6b6cc1530208e007c412ddd062) | diff |
| --- | --- | --- | --- |
| `@kbn/cps-utils` | - | 7 | +7 |
| `@kbn/es-query` | 240 | 241 | +1 |
| `cps` | 6 | 27 | +21 |
| `unifiedSearch` | 119 | 120 | +1 |
| total |  |  | +30 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/8d77e6c618c8d985282dc349a0304ddf631f053d) | [after](https://github.com/elastic/kibana/commit/8f16f442eadaba6b6cc1530208e007c412ddd062) | diff |
| --- | --- | --- | --- |
| `dashboard` | 659.1KB | 659.2KB | +75.0B |
| `unifiedSearch` | 407.9KB | 405.8KB | -2.1KB |
| total |  |  | -2.0KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/8d77e6c618c8d985282dc349a0304ddf631f053d) | [after](https://github.com/elastic/kibana/commit/8f16f442eadaba6b6cc1530208e007c412ddd062) | diff |
| --- | --- | --- | --- |
| `cps` | 1.4KB | 2.5KB | +1.1KB |
| `unifiedSearch` | 22.7KB | 22.8KB | +142.0B |
| total |  |  | +1.2KB |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/8d77e6c618c8d985282dc349a0304ddf631f053d) | [after](https://github.com/elastic/kibana/commit/8f16f442eadaba6b6cc1530208e007c412ddd062) | diff |
| --- | --- | --- | --- |
| `@kbn/cps-utils` | - | 8 | +8 |
| `@kbn/es-query` | 303 | 305 | +2 |
| `cps` | 6 | 27 | +21 |
| `unifiedSearch` | 157 | 158 | +1 |
| total |  |  | +32 |

#### async chunk count
| id | [before](https://github.com/elastic/kibana/commit/8d77e6c618c8d985282dc349a0304ddf631f053d) | [after](https://github.com/elastic/kibana/commit/8f16f442eadaba6b6cc1530208e007c412ddd062) | diff |
| --- | --- | --- | --- |
| `unifiedSearch` | 4 | 5 | +1 |

#### ESLint disabled line counts
| id | [before](https://github.com/elastic/kibana/commit/8d77e6c618c8d985282dc349a0304ddf631f053d) | [after](https://github.com/elastic/kibana/commit/8f16f442eadaba6b6cc1530208e007c412ddd062) | diff |
| --- | --- | --- | --- |
| `@kbn/cps-utils` | - | 1 | +1 |
| `cps` | 4 | 5 | +1 |
| total |  |  | +2 |

#### Total ESLint disabled count
| id | [before](https://github.com/elastic/kibana/commit/8d77e6c618c8d985282dc349a0304ddf631f053d) | [after](https://github.com/elastic/kibana/commit/8f16f442eadaba6b6cc1530208e007c412ddd062) | diff |
| --- | --- | --- | --- |
| `@kbn/cps-utils` | - | 1 | +1 |
| `cps` | 4 | 5 | +1 |
| total |  |  | +2 |


</details>


### History
* :yellow_heart: [Build #357432](https://buildkite.com/elastic/kibana-pull-request/builds/357432) was flaky 970265a1a312dfecde8b403db4d8a593913bd388
* :yellow_heart: [Build #357188](https://buildkite.com/elastic/kibana-pull-request/builds/357188) was flaky 4bb6b4777358dc4487cebf847b030cc4889b72f5

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/358603","number":358603,"commit":"8f16f442eadaba6b6cc1530208e007c412ddd062"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357432","number":357432,"commit":"970265a1a312dfecde8b403db4d8a593913bd388"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/357188","number":357188,"commit":"4bb6b4777358dc4487cebf847b030cc4889b72f5"}],"number":358603}
buildkite-pr-comment-->

TinaHeiligers commented on src/platform/plugins/shared/cps/public/services/cps_manager.ts:
```diff
@@ -0,0 +1,63 @@
+    } catch (error) {
+      // Silently handle errors - keep the current state
+      // eslint-disable-next-line no-console
+      console.error('Failed to fetch projects:', error);
```
Core has a [client-side logging solution](https://github.com/elastic/kibana/pull/144107). Is there any reason you're not using it?

TinaHeiligers reviewed (approved):
LGTM

rudolf commented on src/platform/plugins/shared/cps/public/services/cps_manager.ts:
```diff
@@ -0,0 +1,63 @@
+    } catch (error) {
+      // Silently handle errors - keep the current state
+      // eslint-disable-next-line no-console
+      console.error('Failed to fetch projects:', error);
```
Without any retries this silent failing will result in the UI just appearing to be infinitely loading (projectsSubject never emits)

mbondyra replied:
Actually, the UI wouldn’t display in case of a failure here. But I’ve implemented a retry mechanism to handle this — would that be sufficient or did you have something else in mind?

rudolf reviewed (changes requested):
I don't think we should load the list of project on every page load. I think it will make the CPSManager simpler if it just exposes a `fetchProjects` method instead of a subject and making fetching the responsibility of the component?

mbondyra committed:
Merge commit '2a235bf388e30604ba19ff57bc4234b8f4c1efed' into data-plugin-cps-3

# Conflicts:
#	packages/kbn-optimizer/limits.yml

mbondyra committed:
CPS changes

mbondyra committed:
logger added

mbondyra committed:
limits sorted

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

mbondyra committed:
remove observable since we don't refetch anywhere yet

mbondyra committed:
Merge commit 'f37215ff90c2d7d93dcaae492d862d6b2ce68c01' into data-plugin-cps-3

mbondyra committed:
Merge commit '7518eebd9acaf352f02241901cbeb7129e0a1f32' into data-plugin-cps-3

mbondyra committed:
Merge branch 'main' into data-plugin-cps-3

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/359617)
* Commit: 79319d2154f9215f106f96abbed9b3fb2ba8bdee

### Failed CI Steps
* [Jest Integration Tests #10](https://buildkite.com/elastic/kibana-pull-request/builds/359617#019a6914-4065-4af8-90f9-f868939b30eb)
* [Jest Integration Tests #10](https://buildkite.com/elastic/kibana-pull-request/builds/359617#019a6918-3008-4c0b-987e-d2fa37b4c7df)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/359617","number":359617,"commit":"79319d2154f9215f106f96abbed9b3fb2ba8bdee"}],"number":359617}
buildkite-pr-comment-->

mbondyra committed:
Merge branch 'main' into data-plugin-cps-3

mbondyra committed:
merge server & public type for CPSProject

mbondyra committed:
- relaxing projectRouting type to string
- moving the projectPicker to navigation
- moving types to package to avoid circular dependency, now that the cps plugin mounts component from the package

mbondyra committed:
Merge commit 'de57baa7e8bf50aa094da5f78cc02827d915e239' into data-plugin-cps-3

# Conflicts:
#	src/platform/packages/shared/kbn-es-query/tsconfig.json

kibanamachine committed:
Changes from node scripts/lint_ts_projects --fix

mbondyra committed:
fix types

mbondyra committed:
a comment for cps manager

mbondyra committed:
add presentation to ownership

kibanamachine committed:
Changes from node scripts/generate codeowners

mbondyra committed:
add the notification badge

mbondyra committed:
wip

mbondyra committed:
wip

mbondyra committed:
refactor for better reusability
